### PR TITLE
Ensure we always have called Add() on the inflight counter before we Wait() for it

### PR DESCRIPTION
**Pull request by wvanbergen**
@eapache This is not the cause of #449, but fixes another a potential issue when using the waitgroup.

### CODE REVIEW

File: async_producer.go [hunk 1]
```diff
@@ -610,6 +611,7 @@ func (p *asyncProducer) retryHandler() {
 
 func (p *asyncProducer) shutdown() {
 	Logger.Println("Producer shutting down.")
+	p.inFlight.Add(1)
```
Comment (eapache) [inline async_producer.go]: I don't think this is necessary? If you create a wait group and then immediately Wait on it (without ever calling Add) it just returns right away...

File: async_producer.go [hunk 2]
```diff
@@ -256,7 +257,7 @@ func (p *asyncProducer) topicDispatcher() {
 
 // one per topic
 // partitions messages, then dispatches them by partition
-func (p *asyncProducer) partitionDispatcher(topic string, input chan *ProducerMessage) {
+func (p *asyncProducer) partitionDispatcher(topic string, input <-chan *ProducerMessage) {
```
Comment (eapache) [inline async_producer.go]: these channel restrictions are a good idea though

**Comment by wvanbergen** [comment]
I think, this could potentially be an issue:

``` go
p := NewAsyncProducer(...)
p.Input() <- msg
p.Close()
```

if the `Close` call gets handled before the goroutine has read the first message from the input channel, the inflight counter is still 0.

**Comment by eapache** [comment]
Oh interesting, you're right that's a legit issue. However, I'm not sure your patch fixes all the cases :)
```go
p := NewAsyncProducer(...)
p.Close()
p.Input() <- msg
```
With your patch, if the input message is handled before the waiting goroutine wakes up, then we will increment the counter from 0 before immediately decrementing it... so I think we must stop doing the "increment+decrement" trick for messages we return with the shutting-down error as well.

**Comment by wvanbergen** [comment]
I don't really understand this one? `Close()` doesn't return until the `Wait()` is finished?

**Comment by eapache** [comment]
1. `Close()` calls `Wait()`
2. Go switches contexts
3. the message is sent on `Input()`, is received, causes the wait counter to increment from 0 (the problem case) and decrement again since it is immediately returned as an error

**Comment by wvanbergen** [comment]
(I assume you means `AsyncClose()`, otherwise message on the input channel can never happen until the `Wait()` is finished.)

I don't really see what the problem is: after calling AsyncClose(), no message will ever make it into the processing logic, and we always wait for all the messages being processed to be completed. The worst that can happen is that we wait a bit too long, for `msg` to be returned as error?

**Comment by eapache** [comment]
> I assume you means AsyncClose()

Ya, sorry

> no message will ever make it into the processing logic

But it will still cause the wait counter to be `Add(1)`ed to, which is the bug

**Comment by eapache** [comment]
OK, this is an improvement regardless so I'll merge it; I'll also make a PR which should (I hope) fix and demonstrate the other case I'm thinking of.

**Pull request merged**